README.md
# Проектная работа по DWH для нескольких источников
### Используемые технологии: MongoDB, SQL, Python, Postgres, Airflow, Docker, API

## Конечная задача — реализовать витрину для расчётов с курьерами. 
В ней вам необходимо рассчитать суммы оплаты каждому курьеру за предыдущий месяц. Например, в июне выплачиваем сумму за май. Расчётным числом является 10-е число каждого месяца (это не повлияет на вашу работу). Из витрины бизнес будет брать нужные данные в готовом виде.
Для этого вам нужно создать многослойный DWH и реализовать ETL-процессы, которые будут преобразовывать и перемещать данные от источников до конечных слоёв данных.
Необходимо:
Изучить возможные источники и их внутреннюю модель хранения данных, а также технологии, которые используются для получения данных из этих источников.
На основе требований, целей дальнейшего использования, информации об источниках нужно спроектировать многослойную модель DWH, подробно прорабатывая каждый слой данных.
Для передачи результатов заказчику оформить всю необходимую документацию.

## Описание задачи
Вам необходимо построить витрину, содержащую информацию о выплатах курьерам.
Состав витрины:
id — идентификатор записи.
courier_id — ID курьера, которому перечисляем.
courier_name — Ф. И. О. курьера.
settlement_year — год отчёта.
settlement_month — месяц отчёта, где 1 — январь и 12 — декабрь.
orders_count — количество заказов за период (месяц).
orders_total_sum — общая стоимость заказов.
rate_avg — средний рейтинг курьера по оценкам пользователей.
order_processing_fee — сумма, удержанная компанией за обработку заказов, которая высчитывается как orders_total_sum * 0.25.
courier_order_sum — сумма, которую необходимо перечислить курьеру за доставленные им/ей заказы. За каждый доставленный заказ курьер должен получить некоторую сумму в зависимости от рейтинга (см. ниже).
courier_tips_sum — сумма, которую пользователи оставили курьеру в качестве чаевых.
courier_reward_sum — сумма, которую необходимо перечислить курьеру. Вычисляется как courier_order_sum + courier_tips_sum * 0.95 (5% — комиссия за обработку платежа).

Правила расчёта процента выплаты курьеру в зависимости от рейтинга, где r — это средний рейтинг курьера в расчётном месяце:
r < 4 — 5% от заказа, но не менее 100 р.;
4 <= r < 4.5 — 7% от заказа, но не менее 150 р.;
4.5 <= r < 4.9 — 8% от заказа, но не менее 175 р.;
4.9 <= r — 10% от заказа, но не менее 200 р.
Данные о заказах уже есть в хранилище. Данные курьерской службы вам необходимо забрать из API курьерской службы, после чего совместить их с данными подсистемы заказов.
Отчёт собирается по дате заказа. Если заказ был сделан ночью и даты заказа и доставки не совпадают, в отчёте стоит ориентироваться на дату заказа, а не дату доставки. Иногда заказы, сделанные ночью до 23:59, доставляют на следующий день: дата заказа и доставки не совпадёт. Это важно, потому что такие случаи могут выпадать в том числе и на последний день месяца. Тогда начисление курьеру относите к дате заказа, а не доставки.

## План работ:
1. Изучить API системы доставки заказов
2. Создать таблицы для слоёв в хранилище
3. Реализовать DAGи для всех слоев хранилища





changes.txt
Что было изменено и добавлено в рамках проекта:
1. Добавлен даг payments_to_couriers_dag.py и модуль с основной логикой для дага payments_to_couriers_lib.py
2. Доработаны классы Sale* в рамках модуля stg_to_dds_lib.py с учетом добавления атрибутов доставки и курьера. Добавлены классы для обработки данных о доставке и курьере Delivery* и Courier* соответственно. В даг stg_to_dds_dag.py добавлены таски для прогрузки данных о доставке и курьере.
3. Добавлен даг courier_ledger_dag.py и модуль с основной логикой для дага courier_ledger_lib.py.
4. Добавлены ddl скрипты в /src/ddl/*
5. Добавлено описание атриутов и необходимых изменений /src/api_entities.md

Прочие даги и модули были доработы или реализованы в рамках прохождения всего спринта.


dags/cdm/courier_ledger_dag/courier_ledger_lib.py
from logging import Logger
from typing import List

from cdm.cdm_settings_repository import EtlSetting, DdsEtlSettingsRepository
from lib import PgConnect
from lib.dict_util import json2str
from psycopg import Connection
from psycopg.rows import class_row
from pydantic import BaseModel
from datetime import datetime, date, time

class CourierLedgertObj(BaseModel):
    courier_id: int
    courier_name: str
    settlement_year: int
    settlement_month: int
    orders_count: int
    orders_total_sum: float
    rate_avg: float
    order_processing_fee: float
    courier_order_sum: float
    courier_tips_sum: float
    courier_reward_sum: float
    class Config:
        orm_mode = True


class CLOriginRepository:
    def __init__(self, pg: PgConnect) -> None:
        self._db = pg

    def list_cl(self) -> List[CourierLedgertObj]:
        with self._db.client().cursor(row_factory=class_row(CourierLedgertObj)) as cur:
            cur.execute(
                """
                    with fact as (select fps.courier_id, 
                            dc.courier_name ,
                            dt."year" as settlement_year,
                            dt."month" as settlement_month,
                            count(distinct fps.order_id) as orders_count,
                            sum(fps.total_sum) as orders_total_sum,
                            avg(dd.rate) as rate_avg,
                            case when avg(dd.rate) < 4 then greatest(sum(fps.total_sum) * 0.05, 100)
                                when avg(dd.rate) >= 4 and avg(dd.rate) < 4.5 then greatest(sum(fps.total_sum) * 0.07, 150)
                                when avg(dd.rate) >= 4.5 and avg(dd.rate) < 4.9 then greatest(sum(fps.total_sum) * 0.08, 175)
                                when avg(dd.rate) >= 4.9 then greatest(sum(fps.total_sum) * 0.1, 200) end courier_order_sum,
                            sum(dd.tip_sum) as courier_tips_sum
                    from dds.fct_product_sales fps 
                    inner join dds.dm_couriers dc on dc.id = fps.courier_id 
                    inner join dds.dm_orders do2 on do2.id = fps.order_id 
                    inner join dds.dm_deliveries dd on dd.id = fps.delivery_id 
                    inner join dds.dm_timestamps dt on dt.id  = do2.timestamp_id
                    group by fps.courier_id, 
                            dc.courier_name,
                            settlement_year,
                            settlement_month)
                            
                    select courier_id, 
                            courier_name,
                            settlement_year,
                            settlement_month,
                            orders_count,
                            orders_total_sum,
                            rate_avg,
                            orders_total_sum * 0.25 as order_processing_fee,
                            courier_order_sum,
                            courier_tips_sum,
                            courier_order_sum + courier_tips_sum * 0.95 as courier_reward_sum
                    from fact; 
                """
            )
            objs = cur.fetchall()
        return objs


class CLDestRepository:
    def insert_cl(self, conn: Connection, cl: CourierLedgertObj) -> None:
        with conn.cursor() as cur:
            cur.execute(
                """
                    INSERT INTO cdm.dm_courier_ledger (courier_id, courier_name, settlement_year, settlement_month, orders_count, orders_total_sum, rate_avg, order_processing_fee, courier_order_sum, courier_tips_sum, courier_reward_sum)
                    VALUES (%(courier_id)s, %(courier_name)s, %(settlement_year)s, %(settlement_month)s, %(orders_count)s, %(orders_total_sum)s, %(rate_avg)s, %(order_processing_fee)s, %(courier_order_sum)s, %(courier_tips_sum)s, %(courier_reward_sum)s)
                    ON CONFLICT (courier_id, courier_name, settlement_year, settlement_month) DO UPDATE
                    SET 
                        orders_count = EXCLUDED.orders_count,
                        orders_total_sum = EXCLUDED.orders_total_sum,
                        rate_avg = EXCLUDED.rate_avg,
                        order_processing_fee = EXCLUDED.order_processing_fee,
                        courier_order_sum = EXCLUDED.courier_order_sum,
                        courier_tips_sum = EXCLUDED.courier_tips_sum,
                        courier_reward_sum = EXCLUDED.courier_reward_sum;
                """,
                {
                    "courier_id": cl.courier_id,
                    "courier_name": cl.courier_name,
                    "settlement_year": cl.settlement_year,
                    "settlement_month": cl.settlement_month,
                    "orders_count": cl.orders_count,
                    "orders_total_sum": cl.orders_total_sum,
                    "rate_avg": cl.rate_avg,
                    "order_processing_fee": cl.order_processing_fee,
                    "courier_order_sum": cl.courier_order_sum,
                    "courier_tips_sum": cl.courier_tips_sum,
                    "courier_reward_sum": cl.courier_reward_sum
                },
            )


class CLLoader:
    WF_KEY = "dm_courier_ledger_cdm"
    LAST_LOADED_ID_KEY = "last_loaded_id"
    BATCH_LIMIT = 10000  # Рангов мало, но мы хотим продемонстрировать инкрементальную загрузку рангов.

    def __init__(self, pg_dest: PgConnect, log: Logger) -> None:
        self.pg_dest = pg_dest
        self.origin = CLOriginRepository(pg_dest)
        self.dds = CLDestRepository()
        self.settings_repository = DdsEtlSettingsRepository()
        self.log = log

    def load_cls(self):
        # открываем транзакцию.
        # Транзакция будет закоммичена, если код в блоке with пройдет успешно (т.е. без ошибок).
        # Если возникнет ошибка, произойдет откат изменений (rollback транзакции).
        with self.pg_dest.connection() as conn:
            # Вычитываем очередную пачку объектов.
            load_queue = self.origin.list_cl()
            self.log.info(f"Found {len(load_queue)} users to load.")
            if not load_queue:
                self.log.info("Quitting.")
                return

            # Сохраняем объекты в базу dwh.
            for sr in load_queue:
                self.dds.insert_cl(conn, sr)

            self.log.info(f"Load finished")